### DISCUSSION TITLE
Usage of "signer" in appv2.NDNApp().express()

### DISCUSSION THREAD

**Discussion by mandaou** [named-data/python-ndn#64] [Q&A] [answered]
Hello,
Thank you for all this work.

In reference to [python-ndn](https://github.com/named-data/python-ndn)/[examples](https://github.com/named-data/python-ndn/tree/master/examples)/[appv2](https://github.com/named-data/python-ndn/tree/master/examples/appv2)/basic_packets/  
I'm trying to use app_param (an encoded TlvModel message) and signing interests as per the below:
```python
app = appv2.NDNApp()
keychain = app.default_keychain()

...

        data_name, content, pkt_context = await app.express(
            name, validator=appv2.pass_all, app_param=msg.encode(), signer=keychain.get_signer({}),
            must_be_fresh=True, can_be_prefix=False, lifetime=6000)
``` 
From the consumer side, i am receiving the following error:
```
Sending Interest /example/testApp/randomData/t=1681508953014, InterestParam(can_be_prefix=False, must_be_fresh=True, nonce=None, lifetime=6000, hop_limit=None, forwarding_hint=[])
Timeout

``` 
And from the producer side, I am receiving the follow error:
```
[2023-04-15 00:49:13]WARNING:Drop unvalidated Interest: [<memory at 0x7f5100d247c0>, <memory at 0x7f5100d24dc0>, <memory at 0x7f5100d24e80>, <memory at 0x7f5100d25000>, <memory at 0x7f5100d250c0>]

``` 
And this is the wireshark dump:
![Screenshot from 2023-04-15 00-59-30](https://user-images.githubusercontent.com/95523941/232162109-5d38e40c-c1a2-4705-9a34-8371d8483746.png)

What am I missing here ?

**Comment by zjkmxy** [comment, marked as answer]
Hello.
This is because Appv2 requires you verify all signatures.
On the producer's route registration:
https://github.com/named-data/python-ndn/blob/070e16036b71e001da1c1777beeb942992a8f018/examples/appv2/basic_packets/producer.py#L34-L35
The `validator` argument is by default `None`, which will let the NDNApp drop all signed Interests because the signature is unable to verify.
If you want to pass the validation, you can add `validator=appv2.pass_all` to `route`.

**Comment by mandaou** [comment]
Thank you, adding the validator to the producers' route solved the issue.
